cuqi/distribution/_gaussian.py

```diff
@@ -163,37 +215,342 @@ def gradient(self, val, *args, **kwargs):
+                prec = L_cholmod.inv()
+                sqrtprec = sparse_cholesky(prec)
+                logdet = L_cholmod.logdet()
+                rank = spa.csgraph.structural_rank(cov)# or nplinalg.matrix_rank(cov.todense())
+                # sqrtcov = L_cholmod.L()
+            else:
+                raise NotImplementedError("Sparse standard deviation is only supported via 'cholmod'.")
+                # TODO:
+                # prec = spa.linalg.inv(cov)
+                # sqrtprec = sparse_cholesky(prec)
+                # logdet = np.log(nplinalg.det(cov.todense()))
+                # rank = spa.csgraph.structural_rank(cov)                    
+        else:
+            if dimflag:
+                # this comes from scipy implementation
+                cov = sqrtcov@sqrtcov.T
+                s, u = splinalg.eigh(cov, check_finite=True)
+                eps = eigvalsh_to_eps(s)
+                if np.min(s) < -eps:
+                    raise ValueError("The input matrix must be symmetric positive semidefinite.")                    
+                d = s[s > eps]                    
+                s_pinv = np.array([0 if abs(x) <= eps else 1/x for x in s], dtype=float)
+                
+                U = np.multiply(u, np.sqrt(s_pinv))
+                sqrtprec = U @ np.diag(np.sign(np.diag(U))) #ensure sign is deterministic (scipy gives non-deterministic result)
+                sqrtprec = U.T # We want to have the columns as the eigenvectors
+                
+                rank = len(d)
+                logdet = np.sum(np.log(d))
+                prec = sqrtprec.T @ sqrtprec
+            else:
+                cov = sqrtcov@sqrtcov.T
+                rank = nplinalg.matrix_rank(cov)
+                logdet = np.log(nplinalg.det(cov))
+                prec = nplinalg.inv(cov)
+                sqrtprec = nplinalg.cholesky(prec).T
+    return prec, sqrtprec, logdet, rank
+
+#===================================================
+def get_sqrtprec_from_sqrtprec(dim, sqrtprec, dimflag):
+    # sqrtprec is scalar, corrmat is identity or 1D
+    if (sqrtprec.shape[0] == 1): 
+        logdet = -dim*np.log(sqrtprec**2)
+        rank = dim
+        dia = np.ones(dim)*sqrtprec.flatten()
+        if dimflag:
+            sqrtprec = spa.diags(dia)
+        else:
+            sqrtprec = np.diag(dia)
+    # sqrtprec is vector
+    elif not spa.issparse(sqrtprec) and sqrtprec.shape[0] == np.size(sqrtprec): 
+        logdet = np.sum(-np.log(sqrtprec**2))
+        rank = dim
+        if dimflag:
+            sqrtprec = spa.diags(sqrtprec)
+        else:
+            sqrtprec = np.diag(sqrtprec)
+    # sqrtprec diagonal
+    elif (spa.issparse(sqrtprec) and sqrtprec.format == 'dia') or (not spa.issparse(sqrtprec) and np.count_nonzero(sqrtprec-np.diag(np.diagonal(sqrtprec))) == 0): 
+        precision = sqrtprec@sqrtprec.T
+        logdet = np.sum(-np.log(np.diagonal(precision)))
```

Comment:
This should use precision.diagonal() if the precision is sparse. Also see comment above on the or statement.

Reply:
Done